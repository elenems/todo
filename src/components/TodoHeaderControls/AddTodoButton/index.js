import AddTodoButton from './AddTodoButton.jsx';
export default AddTodoButton;