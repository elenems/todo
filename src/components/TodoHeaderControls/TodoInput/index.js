import TodoInput from './TodoInput.jsx';
export default TodoInput;