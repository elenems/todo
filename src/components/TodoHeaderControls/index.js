import TodoHeaderControls from './TodoHeaderControls.jsx';
export default TodoHeaderControls;