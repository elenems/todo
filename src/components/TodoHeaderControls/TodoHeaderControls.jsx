import React, { Component } from "react";
import TodoInput from "./TodoInput";
import AddTodoButton from "./AddTodoButton";
import { connect } from "react-redux";
import { addTodo, setError } from "../../store/actions/todoActions";
import Box from "@material-ui/core/Box";
import AlertCard from "../../shared/AlertCard";
class TodoHeaderControls extends Component {
  state = {
    title: ""
  };

  changeTitle = e => {
    this.setState({
      [e.target.id]: e.target.value
    });
  };

  addTodo = () => {
    if (this.state.title.length > 0) {
      this.props.addTodo(this.state.title);
      this.setState({
        title: ""
      });
    } else {
      const elements = document.getElementsByClassName("card");
      this.props.setError("New TODO can't be empty!");
      setTimeout(() => {
        for (let elem of elements) {
          elem.className = elem.className.replace(/reveal/, "hide");
        }
        setTimeout(() => {
          this.props.setError("");
        }, 400);
      }, 2500);
      
    }
  };

  render() {
    const { title } = this.state;
    return (
      <div className="controls">
        {this.props.error ? <AlertCard error={this.props.error} /> : null}
        <Box
          display="flex"
          alignItems="baseline"
          justifyContent="space-between"
        >
          <TodoInput value={title} handleChange={this.changeTitle} />
          <AddTodoButton handleClick={this.addTodo} />
        </Box>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    addTodo: title => dispatch(addTodo(title)),
    setError: errorMessage => dispatch(setError(errorMessage))
  };
};

const mapStateToProps = state => {
  return {
    todos: state.todo.todos,
    error: state.todo.error
  };
};

export const TodoHeaderControlsTest = TodoHeaderControls;

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TodoHeaderControls);
