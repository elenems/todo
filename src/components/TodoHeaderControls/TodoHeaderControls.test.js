import { TodoHeaderControlsTest } from "./TodoHeaderControls.jsx";
import { shallow, mount } from "enzyme";
import TodoHeaderControls from "../TodoHeaderControls";
import React from "react";

describe("TodoHeaderControls", () => {
  it("Matches snapshot", () => {
    const wrapper = shallow(<TodoHeaderControlsTest />).debug();
    expect(wrapper).toMatchSnapshot();
  });

  it("Calls changeTitle", () => {
    const wrapper = shallow(<TodoHeaderControlsTest />);
    const mock = jest.spyOn(wrapper.instance(), "changeTitle");
    const event = {
      target: {
        id: 1,
        value: "New title"
      }
    };
    wrapper.instance().changeTitle(event);
    expect(mock).toHaveBeenCalled();
  });

  it("Set state new title", () => {
    const wrapper = shallow(<TodoHeaderControlsTest />);
    wrapper.setState({ title: "New title" });
    expect(wrapper.state("title")).toBe("New title");
  });

  it("Calls setError inside addTodo", () => {
    jest.useFakeTimers();
    const mock = jest.fn();
    const props = {
      setError: title => {
        mock();
      }
    };
    const wrapper = shallow(<TodoHeaderControlsTest {...props} />);
    wrapper.setState({ title: "" });
    wrapper.instance().addTodo();
    expect(mock).toHaveBeenCalled();
    jest.runAllTimers();
    expect(setTimeout).toHaveBeenCalledTimes(2);
    expect(mock).toHaveBeenCalledTimes(2);
  });

  it("Calls addTodo inside addTodo", () => {
    const mock = jest.fn();
    const props = {
      addTodo: () => {
        mock();
      }
    };
    const wrapper = shallow(<TodoHeaderControlsTest {...props} />);
    wrapper.setState({ title: "Some title" });
    wrapper.instance().addTodo();
    expect(mock).toHaveBeenCalled();
    expect(wrapper.state("title")).toBe("");
  });
});
