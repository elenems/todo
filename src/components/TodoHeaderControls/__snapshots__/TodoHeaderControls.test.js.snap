// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoHeaderControls Matches snapshot 1`] = `
"<div className=\\"controls\\">
  <Styled(MuiBox) display=\\"flex\\" alignItems=\\"baseline\\" justifyContent=\\"space-between\\">
    <TodoInput value=\\"\\" handleChange={[Function]} />
    <AddTodoButton handleClick={[Function]} />
  </Styled(MuiBox)>
</div>"
`;
